velox/type/DecimalUtil.h

```diff
@@ -147,6 +148,45 @@ class DecimalUtil {
     return static_cast<TOutput>(rescaledValue);
   }
 
+  /// Rescale a double value to decimal value.
+  ///
+  /// Use `folly::convertTo` to convert a double value to int128_t or int64_t,
+  /// it will throw an exception like 'loss of precision' when the `toValue`
+  /// overflows the limits of int128_t or int64_t. The exception will be caught
+  /// here to judge whether there is a overflow.
+  ///
+  /// Normally, return the rescaled value. Otherwise, if the `toValue` overflows
+  /// the TOutput's limits or the `toValue` exceeds the precision's limits, it
+  /// will throw an exception.
+  template <typename TOutput>
+  inline static std::optional<TOutput> rescaleDouble(
+      const double inputValue,
+      const int toPrecision,
+      const int toScale) {
+    auto toValue =
+        inputValue * static_cast<double>(DecimalUtil::kPowersOfTen[toScale]);
+
+    bool isOverflow = !std::isfinite(toValue);
+    TOutput rescaledValue;
+    try {
+      rescaledValue =
+          util::Converter<CppToType<TOutput>::typeKind>::cast(toValue);
+    } catch (const std::exception& e) {
```

Comment:
Exceptions are super expensive and make try_cast very inefficient. Please, find a better way. See https://github.com/facebookincubator/velox/pull/5801/